### PR TITLE
Implement SQL Subscriber

### PR DESCRIPTION
**Pull request by ChrisBr**
ActiveSupport instrumentation sql.active_record events are getting
now also written into the InfluxDB.

Therefore the RenderSubscriber was splitted into a base class
which gets reused in the SqlSubscriber.

The queries also getting normalized from parameters first, e.g.
'SELECT * FROM posts where id = 42' becomes
'SELECT * FROM posts where id = xxx'. This idea was shameless copied
from http://stevenyue.com/blogs/tracking-sql-queries-in-rails/.

https://guides.rubyonrails.org/active_support_instrumentation.html#sql-active-record

This got extracted #53 and needs to get rebased as soon as #53 is merged.

### CODE REVIEW

**Comment by ChrisBr** [comment]
> I'll sample a few queries from a few different Rails apps I've access to, and throw it against your normalizer. I suspect the interesting queries (those with sub-sejects, joins, and/or group-by clauses) are going to be boiled down too much. I fear that the simple regex is both too simple has a too large CPU footprint (I've seen hundrets of queries on a single page).

That would be great. As said, maybe I can also run it for a few days on our server. With the path element we pretty quickly experienced that it is not usable as a tag.

>> Skylight does something similar, they combine it with the partial and view results

> That looks to me, as if the omit the conditions from the query (I see select from venues, but no select from venue where x in (y). This suggests two tags (table_name and op).

Hm, right. I will maybe dig a little into the skylight gem to see how they normalize the queries.

**Comment by ChrisBr** [comment]
Ok, seems skylight indeed does only store operation and table:

```ruby
require 'skylight'

result = Skylight.lex_sql("SELECT name FROM users WHERE id = 42")
puts result.inspect
# => ["SELECT FROM users", "SELECT name FROM users WHERE id = ?"]
```

Additionally they do the normalization in a C extension. So I agree that we follow your suggestion to store operation and table as tags and the normalized query as a field value.

Question: Do you think the regex will really be a problem performance wise? Do you have any other suggestion (except writing a C extension as well?). Or we just go ahead with the regex for now and implement a switch to disable this metric (e.g. when setting the series_name to nil / ""). What do you think?

**Comment by dmke** [comment]
> Additionally they do the normalization in a C extension. So I agree that we follow your suggestion to store operation and table as tags and the normalized query as a field value.

:+1: 

> Question: Do you think the regex will really be a problem performance wise? Do you have any other suggestion (except writing a C extension as well?). Or we just go ahead with the regex for now and implement a switch to disable this metric (e.g. when setting the series_name to nil / ""). What do you think?

I'd prefer not to write and/or maintain C extensions. Having an off-switch for a regex-based query instrumentation sounds great – but I'd have this feature disabled by default, where the consumer needs to opt-in.

**Comment by ChrisBr** [comment]
> I'd prefer not to write and/or maintain C extensions. Having an off-switch for a regex-based query instrumentation sounds great – but I'd have this feature disabled by default, where the consumer needs to opt-in.

@dmke ok, I will implement this as soon as #53 is implemented.

**Comment by ChrisBr** [comment]
@dmke Rebased and implemented your feedback:

* SQL reports are disabled by default
* We report now sql operation, class name and name
* The normalized query gets stored as an additional value :sql

Please have a look again, thanks :+1:

File: README.md [hunk 1]
```diff
@@ -72,6 +73,10 @@ defined in `lib/influxdb/rails/configuration.rb`
 
 Out of the box, you'll automatically get reporting of your controller,
 view, and db runtimes and rendering of template, partial and collection for each request.
+Reporting of SQL queries is disabled by default because it is still in experimental mode
+and currently requires String parsing which might cause performance issues on query
+intensive applications. You can enable it by setting the ``series_name_for_sql``
```
Comment (dmke) [inline README.md]:
```diff
- ``series_name_for_sql``
+ `series_name_for_sql`
```

File: lib/influxdb/rails/middleware/request_subscriber.rb [hunk 2]
```diff
@@ -8,14 +8,12 @@ def call(_name, start, finish, _id, payload) # rubocop:disable Metrics/MethodLen
...
           begin
             series(payload, start, finish).each do |series_name, value|
-              InfluxDB::Rails.client.write_point \
-                series_name,
-                values:    { value: value },
-                tags:      tags,
-                timestamp: ts
+              InfluxDB::Rails.client.write_point series_name,
+                                                 values:    { value: value },
+                                                 tags:      tags(payload),
```
Comment (dmke) [inline lib/influxdb/rails/middleware/request_subscriber.rb]: Please keep the optimization introduced in 1734e4fac98708a92511a1f97084daf232e13cfb.

File: spec/unit/sql/query.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  describe "#class_name" do
+    it { expect(subject.class_name).to eq("users") }
+  end
+
+  describe "#operation" do
+    it { expect(subject.operation).to eq("SELECT") }
+  end
+
+  describe "#track?" do
+    it { expect(described_class.new("INSERT").table_name).to be true }
```
Comment (dmke) [inline spec/unit/sql/query.rb]: This seems to confuse `track?` with `table_name`.

File: lib/influxdb/rails/middleware/simple_subscriber.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+          end
+        end
+
+        private
+
+        def values(started, finished, _payload)
+          { value: ((finished - started) * 1000).ceil }
+        end
+
+        def time_stamp(finished)
```
Comment (dmke) [inline lib/influxdb/rails/middleware/simple_subscriber.rb]: Please use `timestamp`, not `time_stamp`.

**Review by dmke — CHANGES_REQUESTED** [review]
Again, some nitpicks. I'm not sure whether the `SimpleSubscriber` class add an unnecessary level of abstraction... let me think about this a bit.

File: spec/unit/sql/query.rb (hunk 3 above)
Comment (ChrisBr) [reply spec/unit/sql/query.rb]: You're right ... and it was not even executed because of the missing `_spec.rb` postfix :man_facepalming:

**Comment by ChrisBr** [comment]
@dmke fixed

> I'm not sure whether the SimpleSubscriber class add an unnecessary level of abstraction... let me think about this a bit.

Hm, I see you ... however, the render and sql subscriber share quite some functionality :thinking: I was even thinking about moving everything into the Subscriber class and just overwrite the call method in ``RequestSubscriber`` as this potentially is the only subscriber which sends several datapoints to Influx. All other subscriber send a single data point. Or maybe we also split the request subscriber into just sending a single datapoint but subscribe it three times instead? Might be a little cleaner ...

**Comment by dmke** [comment]
> `RequestSubscriber` [...] is the only subscriber which sends several datapoints to Influx.
> [...] split the request subscriber into just sending a single datapoint but subscribe it three times instead? Might be a little cleaner ...

Cleaner maybe, but this would be unnecessary implementation work for a temporary workaround. Pending #56, I'd like to resolve this in the next week, so you don't need to touch this ;-)

File: spec/unit/sql/query.rb (hunk 3 above)
Comment (dmke) [reply spec/unit/sql/query.rb]: A shame Rubocop didn't catch that, whilst it complains about *everything* else :-D

Comment (ChrisBr) [reply spec/unit/sql/query.rb]: image


https://github.com/rubocop-hq/rubocop-rspec

Comment (dmke) [reply spec/unit/sql/query.rb]: That... causes too much work :-)

**Comment by ChrisBr** [comment]
Thanks for merging and releasing a new version :heart: And happy birthday (xing...) :cake: :tada:

**Comment by dmke** [comment]
No problem (and thanks! 🎈)

**Pull request merged**